### PR TITLE
Add Support for CBC, CTR Cipher modes with AES 128 & 256 bit keys.

### PR DESCRIPTION
**Pull request by skmcgrail**
### Description of changes: 
This pull request adds support for CBC and CTR cipher mode encryption for 128-bit and 256-bit AES keys.

* Introduces `aws_lc_rs::cipher` module which supports unauthenticated ciphers for encryption/decryption.
* `PaddedBlockEncryptingKey`, `PaddedBlockDecryptingKey` provide encryption/decryption of plaintext/ciphertext for block cipher modes using padding. `PaddedBlockEncryptingKey::cbc_pkcs7(...)` `PaddedBlockDecryptingKey::cbc_pkcs7(...)` and are the constructor methods provided for using CBC mode with PKCS#7 padding. 
  * Users who need to explicitly set an initialization vector can use `PaddedBlockEncryptingKey::less_safe_cbc_pkcs7`.
* `EncryptingKey`, `DecryptingKey` provide encryption/decryption of plaintext/ciphertext for unpadded block cipher modes. `PaddedBlockEncryptingKey::ctr(...)` `PaddedBlockDecryptingKey::ctr(...)` and are the constructor methods provided for using CTR mode (unpadded). 
  * Users who need to explicitly set an initialization vector can use `EncryptingKey::less_safe_ctr`.

#### AES-128 CBC
```rust
use aws_lc_rs::cipher::{
    PaddedBlockDecryptingKey, PaddedBlockEncryptingKey, UnboundCipherKey, AES_128,
};

let original_message = "This is a secret message!".as_bytes();
let mut in_out_buffer = Vec::from(original_message);

let key_bytes: &[u8] = &[
    0xff, 0x0b, 0xe5, 0x84, 0x64, 0x0b, 0x00, 0xc8, 0x90, 0x7a, 0x4b, 0xbf, 0x82, 0x7c, 0xb6,
    0xd1,
];

let key = UnboundCipherKey::new(&AES_128, key_bytes)?;
let encrypting_key = PaddedBlockEncryptingKey::cbc_pkcs7(key)?;
let context = encrypting_key.encrypt(&mut in_out_buffer)?;

let key = UnboundCipherKey::new(&AES_128, key_bytes)?;
let decrypting_key = PaddedBlockDecryptingKey::cbc_pkcs7(key)?;
let plaintext = decrypting_key.decrypt(&mut in_out_buffer, context)?;
assert_eq!(original_message, plaintext);
```

#### AES-128 CTR
```rust
use aws_lc_rs::cipher::{DecryptingKey, EncryptingKey, UnboundCipherKey, AES_128};

let original_message = "This is a secret message!".as_bytes();
let mut in_out_buffer = Vec::from(original_message);

let key_bytes: &[u8] = &[
    0xff, 0x0b, 0xe5, 0x84, 0x64, 0x0b, 0x00, 0xc8, 0x90, 0x7a, 0x4b, 0xbf, 0x82, 0x7c, 0xb6,
    0xd1,
];

let key = UnboundCipherKey::new(&AES_128, key_bytes)?;
let encrypting_key = EncryptingKey::ctr(key)?;
let context = encrypting_key.encrypt(&mut in_out_buffer)?;

let key = UnboundCipherKey::new(&AES_128, key_bytes)?;
let decrypting_key = DecryptingKey::ctr(key)?;
let plaintext = decrypting_key.decrypt(&mut in_out_buffer, context)?;
assert_eq!(original_message, plaintext);
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

### CODE REVIEW

**Review by samuel40791765 — COMMENTED** [review]
First pass; I'm a bit curious what the benchmark numbers look like against `openssl_bench` :)

File: aws-lc-rs/src/cipher.rs [hunk 1]
```diff
@@ -3,158 +3,1268 @@
...
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> Result<(), core::fmt::Error> {
+        f.debug_struct("UnboundCipherKey")
+            .field("algorithm", &self.algorithm)
+            .finish()
+    }
+}
+
+impl From<hkdf::Okm<'_, &'static Algorithm>> for UnboundCipherKey {
+    fn from(okm: hkdf::Okm<&'static Algorithm>) -> Self {
+        let mut key_bytes = [0; AES_256_KEY_LEN];
```
[the remark this answers is not in the document]
Comment (justsmth) [reply aws-lc-rs/src/cipher.rs]: I agree. Latest commit uses a `MAX_CIPHER_KEY_LEN`.

File: aws-lc-rs/src/cipher.rs [hunk 2]
```diff
@@ -3,158 +3,1268 @@
...
+                // This implements PKCS#7 padding scheme, used by aws-lc if we were using EVP_CIPHER API's
+                let remainder = in_out_len % block_len;
+                if remainder == 0 {
+                    let block_size: u8 = block_len.try_into().map_err(|_| Unspecified)?;
+                    in_out.extend(vec![block_size; block_len].iter());
+                } else {
+                    let padding_size = block_len - remainder;
+                    let v: u8 = padding_size.try_into().map_err(|_| Unspecified)?;
+                    // Heap allocation :(
+                    in_out.extend(vec![v; padding_size].iter());
```
[the remark this answers is not in the document]
Comment (justsmth) [reply aws-lc-rs/src/cipher.rs]: It would apply to both branches. 

I updated the implementation in the latest commit to use a 16-byte `padding_buffer` instead of the `vec!` macro. With this change it's possibly not a "heap allocation" depending on the "capacity" of the collection and how it implements the [`Extend`](https://dev-doc.rust-lang.org/beta/std/iter/trait.Extend.html) trait.

File: aws-lc-rs/src/cipher/key.rs [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+            SymmetricCipherKey::Aes128 { enc_key, .. }
+            | SymmetricCipherKey::Aes256 { enc_key, .. } => encrypt_block_aes(enc_key, block),
+            SymmetricCipherKey::ChaCha20 { .. } => panic!("Unsupported algorithm!"),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::cipher::block::{Block, BLOCK_LEN};
```
Comment (samuel40791765) [inline aws-lc-rs/src/cipher/key.rs:170]: Is `BLOCK_LEN` redundant with `AES_BLOCK_LEN`, or do they imply different things?

File: aws-lc-rs/examples/cipher.rs [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+
+fn aes_ctr_encrypt(key: &[u8], iv: Option<String>, plaintext: String) -> Result<(), &'static str> {
+    let hex_key = hex::encode(key);
+    let key = new_unbound_key(key)?;
+
+    let key = match iv {
+        Some(iv) => {
+            let iv = {
+                let v = hex::decode(iv).map_err(|_| "invalid iv")?;
+                let v: FixedLength<16> = v.as_slice().try_into().map_err(|_| "invalid iv")?;
```
Comment (samuel40791765) [inline aws-lc-rs/examples/cipher.rs:107]: Can we use `AES_IV_LEN` here? Same for the other instances below, it might be best to show in our examples that the user can depend on the predefined values we provide.
```suggestion
                let v: FixedLength<AES_IV_LEN> = v.as_slice().try_into().map_err(|_| "invalid iv")?;
```

File: aws-lc-rs/src/cipher/key.rs [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+            Ok(SymmetricCipherKey::Aes128 {
+                raw_key: Aes128Key(kb.assume_init()),
+                enc_key,
+                dec_key,
+            })
+        }
+    }
+
+    pub(crate) fn aes256(key_bytes: &[u8]) -> Result<Self, Unspecified> {
+        if key_bytes.len() != 32 {
```
Comment (samuel40791765) [inline aws-lc-rs/src/cipher/key.rs:99]:
```suggestion
        if key_bytes.len() != AES_256_KEY_LEN {
```
Maybe we can depend on `AES_128_KEY_LEN` and `AES_256_KEY_LEN` for all the instances of `16` and `32` in this file (except for `chacha20`).

**Review by samuel40791765 — COMMENTED** [review]
initialization in initialization vector seems to be misspelled throughout the PR, so we might want to adjust that.

File: aws-lc-rs/src/cipher.rs [hunk 6]
```diff
@@ -3,158 +3,1270 @@
...
-                value.zeroize();
+            Self::Iv128(_) => write!(f, "Iv128"),
+            Self::None => write!(f, "None"),
+        }
+    }
+}
+
+#[non_exhaustive]
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+/// Cipher algorithm identifier.
```
Comment (torben-hansen) [inline aws-lc-rs/src/cipher.rs:283]: should we call this "block cipher", which is the standard nomenclature. But I guess all those names are just a subset of "algorithm" anyway.

File: aws-lc-rs/src/cipher.rs [hunk 7]
```diff
@@ -3,158 +3,1270 @@
...
+    PKCS7,
+}
+
+impl PaddingStrategy {
+    fn add_padding<InOut>(self, block_len: usize, in_out: &mut InOut) -> Result<(), Unspecified>
+    where
+        InOut: AsMut<[u8]> + for<'in_out> Extend<&'in_out u8>,
+    {
+        match self {
+            PaddingStrategy::PKCS7 => {
```
Comment (torben-hansen) [inline aws-lc-rs/src/cipher.rs:116]: one restriction is that the padding length must be strictly less than 256 (otherwise it doesn't fit in an octet).

File: aws-lc-rs/src/cipher.rs [hunk 8]
```diff
@@ -3,158 +3,1270 @@
...
+        EncryptingKey::new(key, OperatingMode::CTR, None)
+    }
+
+    /// Constructs an `EncryptingKey` operating in counter (CTR) mode using the provided key.
+    /// The users provided context will be used for the CTR mode initalization-vector.
+    ///
+    /// # Errors
+    /// * [`Unspecified`]: Returned if there is an error creating the `EncryptingKey`.
+    ///
+    pub fn less_safe_ctr(
```
Comment (torben-hansen) [inline aws-lc-rs/src/cipher.rs:621]: Still pondering on the name. It's not inherently less safe than the other function call....

File: aws-lc-rs/src/cipher.rs [hunk 9]
```diff
@@ -3,158 +3,1270 @@
...
+    pub fn mode(&self) -> OperatingMode {
+        self.mode
+    }
+
+    /// Decrypts the data provided in `in_out` in-place.
+    /// Returns a references to the decrypted data.
+    ///
+    /// # Errors
+    /// * [`Unspecified`]: Returned if cipher mode requires input to be a multiple of the block length,
+    /// and `in_out.len()` is not. Otherwise returned if decryption fails.
```
Comment (torben-hansen) [inline aws-lc-rs/src/cipher.rs:751]:
```suggestion
    /// and `in_out.len()` is not. Also returned if decryption fails.
```
?

File: aws-lc-rs/src/cipher/key.rs (hunk 3 above)
Comment (justsmth) [reply aws-lc-rs/src/cipher/key.rs:170]: Both the type `Block` and the value `BLOCK_LEN` are only internally visible (`pub(crate)`). The type is primarily used for [`aead::HeaderProtectionKey::new_mask`](https://docs.rs/aws-lc-rs/latest/aws_lc_rs/aead/quic/struct.HeaderProtectionKey.html#method.new_mask) which uses a "sample_len" of 16 (even for ChaCha20), which also happens to align with `BLOCK_LEN`.

File: aws-lc-rs/src/cipher.rs [hunk 10]
```diff
@@ -3,158 +3,1270 @@
...
 use zeroize::Zeroize;
 
-pub(crate) enum SymmetricCipherKey {
-    Aes128(Aes128Key, AES_KEY),
-    Aes256(Aes256Key, AES_KEY),
-    ChaCha20(ChaCha20Key),
+/// The cipher block padding strategy.
+#[non_exhaustive]
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum PaddingStrategy {
```
[the remark this answers is not in the document]
Comment (justsmth) [reply aws-lc-rs/src/cipher.rs:105]: * Having a method that only provides a specific *mode* and *padding* combination is intentional. The goal is guide the user toward the most common (or the "best practice" for) mode/padding combinations. We anticipate adding methods for other combinations, or even opening it up to arbitrary combinations in the future.
* I prefer to expose this type to consumers, but I agree that an enum with only one variant is not very useful.

File: aws-lc-rs/src/cipher.rs [hunk 11]
```diff
@@ -3,158 +3,1270 @@
...
+        }
+
+        let cipher_key = UnboundCipherKey::new(alg, key).unwrap();
+        let encrypting_key = EncryptingKey::new(cipher_key, mode, None).unwrap();
+
+        let mut in_out = input.clone();
+        let decrypt_iv = encrypting_key.encrypt(&mut in_out).unwrap();
+
+        if n > 5 {
+            // There's no more than a 1 in 2^48 chance that this will fail randomly
```
[the remark this answers is not in the document]
Comment (justsmth) [reply aws-lc-rs/src/cipher.rs:993]: This is a test where we originally didn't have the `n > 5` condition, so with 1 byte there was about a (1/256) probability that the ciphertext would be the same as the plaintext and the next test condition would fail.  For n bytes, the probability of a match is around 1/(2^(n*8)).

File: aws-lc-rs/src/cipher.rs (hunk 6 above)
Comment (justsmth) [reply aws-lc-rs/src/cipher.rs:283]: We intend for this API to potentially support Stream ciphers (like ChaCha) as well.

**Comment by skmcgrail** [comment]
https://github.com/awslabs/aws-lc-rs/pull/155 has some API improvements which will allow reuse of the Encrypt/Decrypt keys so that they aren't consumed by value after calling encrypt/decrypt. Will allow us a path for future optimizations if/when we decide to migrate to the EVP_CIPHER based APIs.

File: aws-lc-rs/src/cipher/key.rs [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+    }
+}
+
+impl SymmetricCipherKey {
+    pub(crate) fn aes128(key_bytes: &[u8]) -> Result<Self, Unspecified> {
+        if key_bytes.len() != AES_128_KEY_LEN {
+            return Err(Unspecified);
+        }
+
+        unsafe {
```
[the remark this answers is not in the document]
Comment (camshaft) [reply aws-lc-rs/src/cipher/key.rs:64]: If I'm not mistaken, this is the same for the [AEAD types](https://docs.rs/aws-lc-rs/1.1.0/aws_lc_rs/aead/index.html) as well. I wish it wasn't like this, though, since we have to pay both the latency and memory costs. This is especially wasteful for things like TLS and QUIC where we have separate derived sealer and opener keys for both directions so the other halves never get used.

File: aws-lc-rs/src/hkdf.rs [hunk 13]
```diff
@@ -92,6 +92,7 @@ pub struct Salt {
     key_len: usize,
 }
 
+#[allow(clippy::missing_fields_in_debug)]
```
Comment (camshaft) [inline aws-lc-rs/src/hkdf.rs:95]: nit: This seems unrelated to this PR. In the future, it would be easier to review with more focused PRs.

File: aws-lc-rs/src/iv.rs [hunk 14]
```diff
@@ -0,0 +1,98 @@
+// Copyright 2018 Brian Smith.
+// SPDX-License-Identifier: ISC
+// Modifications copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+#![allow(dead_code)]
+
+//! Initialization Vector (IV) cryptographic primitives
```
Comment (camshaft) [inline aws-lc-rs/src/iv.rs:7]: This module could also be a separate PR, along with using it everywhere that isn't the CBC/CTR stuff

File: aws-lc-rs/tests/cipher_test.rs [hunk 15]
```diff
@@ -0,0 +1,360 @@
...
+            assert_eq!(expected_ciphertext.as_slice(), in_out.as_slice());
+
+            let unbound_key2 = UnboundCipherKey::new($alg, &key).unwrap();
+            let decrypting_key =
+                PaddedBlockDecryptingKey::$constructor(unbound_key2, context).unwrap();
+            assert_eq!($mode, decrypting_key.mode());
+            assert_eq!($padding, decrypting_key.padding());
+            assert_eq!($alg, decrypting_key.algorithm());
+            let plaintext = decrypting_key.decrypt(&mut in_out).unwrap();
+            assert_eq!(input.as_slice(), plaintext);
```
Comment (camshaft) [inline aws-lc-rs/tests/cipher_test.rs:41]: nit: it might be worth extracting some of this out into an actual function rather than a macro. It's especially annoying to work with `paste!` and other tooling like rustfmt or rust-analyzer.

**Review by camshaft — COMMENTED** [review]
I got through everything outside of `cipher.rs` and `cipher/key.rs`. I'll need to do the rest tomorrow.

File: aws-lc-rs/src/cipher/key.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (torben-hansen) [reply aws-lc-rs/src/cipher/key.rs:64]: All uni-directional protocols will have to pay this cost. Probably somewhat minor(?), but still.

File: aws-lc-rs/benches/cipher_benchmark.rs [hunk 16]
```diff
@@ -0,0 +1,165 @@
+use aws_lc_rs::cipher::{
```
[the remark this answers is not in the document]
Comment (justsmth) [reply aws-lc-rs/benches/cipher_benchmark.rs:1]: PR: https://github.com/awslabs/aws-lc-rs/pull/158

File: aws-lc-rs/src/cipher/key.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (justsmth) [reply aws-lc-rs/src/cipher/key.rs:64]: We could lazily initialize these values at the cost of some code complexity and the additional latency of a branch or two in the logic. These are optimizations we should make, but in a separate PR.

On a related note, my [PR here](https://github.com/awslabs/aws-lc-rs/pull/156) eliminates one of the fields of this struct.

File: aws-lc-rs/src/cipher.rs [hunk 17]
```diff
@@ -3,158 +3,1270 @@
...
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum OperatingMode {
+    /// Cipher block chaining (CBC) mode.
+    CBC,
+
+    /// Counter (CTR) mode.
+    CTR,
+}
+
+/// The contextual data used to encrypted/decrypt data.
```
[the remark this answers is not in the document]
Comment (skmcgrail) [reply aws-lc-rs/src/cipher.rs:189]: Examples in the repository aren't immediately discoverable unless the user looks for them, or you direct them to a hard-coded link to them which might change later. This allows the code to be directly inlined into the documentation, and it is tested as part of `cargo test` ensuring it doesn't drift in the future.

File: aws-lc-rs/src/cipher.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (samuel40791765) [reply aws-lc-rs/src/cipher.rs]: Yeah I like the concept of exposing it once we have support for additional variants.

File: aws-lc-rs/src/hkdf.rs (hunk 13 above)
Comment (justsmth) [reply aws-lc-rs/src/hkdf.rs]: The latest commit merged changes (like this one) which were already on main, so some of the "noise" like this should be gone from the diff.

**Review by torben-hansen — APPROVED** [review]
CI looks unhappy. But otherwise, OK from me.

**Comment by justsmth** [comment]
> CI looks unhappy. But otherwise, OK from me.

CI is happy now.

**Review by samuel40791765 — APPROVED** [review]
